app/services/trainees/create_from_dttp.rb

```diff
@@ -0,0 +1,333 @@
+        last_name: dttp_trainee.response["lastname"],
+        date_of_birth: dttp_trainee.date_of_birth,
+        gender: trainee_gender,
+        nationalities: nationalities,
+      }
+    end
+
+    def locale_code
+      return {} if dttp_trainee.response["address1_line1"].blank?
+
+      { locale_code: Trainee.locale_codes[:uk] }
+    end
+
+    def contact_attributes
+      {
+        address_line_one: dttp_trainee.response["address1_line1"],
+        address_line_two: dttp_trainee.response["address1_line2"],
+        town_city: dttp_trainee.response["address1_line3"],
+        postcode: dttp_trainee.response["address1_postalcode"],
+        email: dttp_trainee.response["emailaddress1"],
+      }.merge(locale_code)
+    end
+
+    def course_attributes
+      {
+        course_subject_one: course(latest_placement_assignment.response["_dfe_ittsubject1id_value"]),
+        course_subject_two: course(latest_placement_assignment.response["_dfe_ittsubject2id_value"]),
+        course_subject_three: course(latest_placement_assignment.response["_dfe_ittsubject3id_value"]),
+        course_min_age: age_range && age_range[0],
+        course_max_age: age_range && age_range[1],
+        study_mode: study_mode,
+        commencement_date: earliest_placement_assignment.response["dfe_commencementdate"],
+        course_start_date: latest_placement_assignment.response["dfe_programmestartdate"],
+        course_end_date: latest_placement_assignment.response["dfe_programmeeenddate"],
+      }
+    end
+
+    def course(dttp_course_uuid)
+      find_by_entity_id(
+        dttp_course_uuid,
+        Dttp::CodeSets::CourseSubjects::MAPPING,
+      )
+    end
+
+    def age_range
+      @age_range ||= find_by_entity_id(
+        latest_placement_assignment.response["_dfe_coursephaseid_value"],
+        Dttp::CodeSets::AgeRanges::MAPPING,
+      )
+    end
+
+    def study_mode
+      find_by_entity_id(
+        latest_placement_assignment.response["_dfe_studymodeid_value"],
+        Dttp::CodeSets::CourseStudyModes::MAPPING,
+      )
+    end
+
+    def school_attributes
+      return {} if latest_placement_assignment.lead_school_id.blank?
+
+      # Should we raise when schools are not found so that we can add them?
+      attrs = {
+        lead_school: School.find_by(urn: lead_school_urn),
+      }
+
+      if latest_placement_assignment.employing_school_id.present?
+        attrs.merge!({
+          employing_school: School.find_by(urn: employing_school_urn),
+        })
+      end
+
+      attrs
+    end
+
+    def lead_school_urn
+      Dttp::School.find_by(dttp_id: latest_placement_assignment.lead_school_id)&.urn
+    end
+
+    def employing_school_urn
+      Dttp::School.find_by(dttp_id: latest_placement_assignment.employing_school_id)&.urn
+    end
+
+    def funding_attributes
+      return {} unless latest_placement_assignment.response["dfe_allocatedplace"] == Dttp::Params::PlacementAssignment::ALLOCATED_PLACE
```

Comment:
I'm not 100% sure about this field... are there any dttp_trainees with `_dfe_bursarydetailsid_value` not null but `dfe_allocatedplace` null?

Reply:
I've reverse mapped it from the snippet below, but happy to change if my assumption is wrong.
https://github.com/DFE-Digital/register-trainee-teachers/blob/e5a50a747ef7f04b6db33919d09f1ff3c4c28b9c/app/lib/dttp/params/placement_assignment.rb#L157-L166

Reply:
> I'm not 100% sure about this field... are there any dttp_trainees with `_dfe_bursarydetailsid_value` not null but `dfe_allocatedplace` null?

I've changed this now so that only an explicit `Dttp::Params::
PlacementAssignment::NO_ALLOCATED_PLACE` will cause funding to be not set.